federation/pkg/federation-controller/util/federated_informer_test.go

```diff
@@ -0,0 +1,126 @@
+limitations under the License.
+*/
+
+package util
+
+import (
+	"testing"
+	"time"
+
+	federation_api "k8s.io/kubernetes/federation/apis/federation/v1beta1"
+	federation_release_1_4 "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	fake_federation_release_1_4 "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4/fake"
+	api "k8s.io/kubernetes/pkg/api"
+	api_v1 "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/testing/core"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Basic test for Federated Informer. Checks whether the subinformer are added and deleted
+// when the corresponding cluster entries appear and dissapear from etcd.
+func TestFederatedInformer(t *testing.T) {
+	fakeClient := &fake_federation_release_1_4.Clientset{}
+
+	// Add a single cluster to federation and remove it when needed.
+	cluster := federation_api.Cluster{
+		ObjectMeta: api_v1.ObjectMeta{
+			Name: "mycluster",
+		},
+		Status: federation_api.ClusterStatus{
+			Conditions: []federation_api.ClusterCondition{
+				{Type: federation_api.ClusterReady, Status: api_v1.ConditionTrue},
+			},
+		},
+	}
+	fakeClient.AddReactor("list", "clusters", func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		return true, &federation_api.ClusterList{Items: []federation_api.Cluster{cluster}}, nil
+	})
+	deleteChan := make(chan struct{})
+	fakeClient.AddWatchReactor("clusters", func(action core.Action) (handled bool, ret watch.Interface, err error) {
+		fakeWatch := watch.NewFake()
+		go func() {
+			<-deleteChan
+			fakeWatch.Delete(&cluster)
+		}()
+		return true, fakeWatch, nil
+	})
+
+	// There is a single service ns1/s1 in cluster mycluster.
+	service := api_v1.Service{
+		ObjectMeta: api_v1.ObjectMeta{
+			Namespace: "ns1",
+			Name:      "s1",
+		},
+	}
+	fakeClient.AddReactor("list", "services", func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		return true, &api_v1.ServiceList{Items: []api_v1.Service{service}}, nil
+	})
+	fakeClient.AddWatchReactor("services", func(action core.Action) (handled bool, ret watch.Interface, err error) {
+		return true, watch.NewFake(), nil
+	})
+
+	targetInformerFactory := func(cluster *federation_api.Cluster, clientset federation_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+		return framework.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+					return clientset.Core().Services(api_v1.NamespaceAll).List(options)
+				},
+				WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+					return clientset.Core().Services(api_v1.NamespaceAll).Watch(options)
+				},
+			},
+			&api_v1.Service{},
+			10*time.Second,
+			framework.ResourceEventHandlerFuncs{
```

Comment:
framework.ResourceEventHandlerFuncs{}

(since you are not adding any handler functions)